flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SnapshotObserver.java

```diff
@@ -219,64 +240,205 @@ private void observeTriggeredCheckpoint(FlinkResourceContext<CR> ctx, String job
+                        .collect(Collectors.toList());
+
+        var lastCompleteSnapshot =
+                snapshotList.stream()
+                        .filter(s -> COMPLETED.equals(s.getStatus().getState()))
+                        .max(Comparator.comparing(EXTRACT_SNAPSHOT_TIME))
+                        .orElse(null);
+
+        var maxCount = getMaxCountForSnapshotType(observeConfig, operatorConfig, snapshotType);
+        var maxTms = getMinAgeForSnapshotType(observeConfig, operatorConfig, snapshotType);
+        var result = new HashSet<FlinkStateSnapshot>();
+
+        if (snapshotList.size() < 2) {
+            return result;
+        }
+
+        for (var snapshot : snapshotList) {
+            if (snapshot.equals(lastCompleteSnapshot)) {
+                continue;
+            }
+
+            var ts = EXTRACT_SNAPSHOT_TIME.apply(snapshot).toEpochMilli();
+            if (snapshotList.size() - result.size() > maxCount || ts < maxTms) {
+                result.add(snapshot);
+            }
+        }
 
-        var observeConfig = ctx.getObserveConfig();
-        var flinkService = ctx.getFlinkService();
-        boolean savepointCleanupEnabled =
-                observeConfig.getBoolean(
-                        KubernetesOperatorConfigOptions.OPERATOR_SAVEPOINT_CLEANUP_ENABLED);
+        return result;
+    }
 
-        // maintain history
-        List<Savepoint> savepointHistory = currentSavepointInfo.getSavepointHistory();
-        if (savepointHistory.size() < 2) {
+    /**
+     * Cleans up the savepoint history of a Flink resource from the old, deprecated
+     * savepoint-history. Secondary resources of FlinkStateSnapshot are used to determine count of
+     * completed savepoints to be able to properly clean old savepoints based on the count policy.
+     *
+     * @param ctx flink resource context
+     * @param allSecondarySnapshotResources all snapshot resources linked to this Flink resource
+     */
+    @VisibleForTesting
+    void cleanupSavepointHistoryLegacy(
+            FlinkResourceContext<CR> ctx, Set<FlinkStateSnapshot> allSecondarySnapshotResources) {
+        var maxTms =
+                getMinAgeForSnapshotType(
+                        ctx.getObserveConfig(), ctx.getOperatorConfig(), SAVEPOINT);
+        var maxCount =
+                getMaxCountForSnapshotType(
+                        ctx.getObserveConfig(), ctx.getOperatorConfig(), SAVEPOINT);
+
+        var completedSavepointCrs =
+                allSecondarySnapshotResources.stream()
+                        .filter(
+                                s ->
+                                        s.getStatus() != null
+                                                && COMPLETED.equals(s.getStatus().getState()))
+                        .filter(s -> s.getSpec().isSavepoint())
+                        .count();
+        maxCount = Math.max(0, maxCount - completedSavepointCrs);
```

Comment:
Shouldn't we always keep at least 1?

Reply:
I have added another `Math.max` to `getMaxCountForSnapshotType`. Here we can use 0, as we should be able to clear the list of legacy savepoints if we have enough completed FlinkStateSnapshot resources.